Sources/SourceKitLSP/Rename.swift

```diff
@@ -268,33 +254,146 @@ fileprivate struct SyntacticRenameName {
+    }
+
+    if let usr = renameResult.usr, let oldName = renameResult.oldName, let index = workspace.index {
+      // If we have a USR + old name, perform an index lookup to find workspace-wide symbols to rename.
+      // First, group all occurrences of that USR by the files they occur in.
+      var locationsByFile: [URL: [RenameLocation]] = [:]
+      let occurrences = index.occurrences(ofUSR: usr, roles: [.declaration, .definition, .reference])
+      for occurrence in occurrences {
+        let url = URL(fileURLWithPath: occurrence.location.path)
+        let renameLocation = RenameLocation(
+          line: occurrence.location.line,
+          utf8Column: occurrence.location.utf8Column,
+          usage: RenameLocation.Usage(roles: occurrence.roles)
+        )
+        locationsByFile[url, default: []].append(renameLocation)
+      }
+
+      // Now, call `editsToRename(locations:in:oldName:newName:)` on the language service to convert these ranges into
+      // edits.
+      await withTaskGroup(of: (DocumentURI, [TextEdit])?.self) { taskGroup in
+        for (url, renameLocations) in locationsByFile {
+          let uri = DocumentURI(url)
+          if edits.changes![uri] != nil {
+            // We already have edits for this document provided by the language service, so we don't need to compute
+            // rename ranges for it.
+            continue
+          }
+          taskGroup.addTask {
+            // Create a document snapshot to operate on. If the document is open, load it from the document manager,
+            // otherwise conjure one from the file on disk. We need the file in memory to perform UTF-8 to UTF-16 column
+            // conversions.
+            // We should technically infer the language for the from-disk snapshot. But `editsToRename` doesn't care
+            // about it, so defaulting to Swift is good enough for now
+            // If we fail to get edits for one file, log an error and continue but don't fail rename completely.
+            guard
+              let snapshot = (try? self.documentManager.latestSnapshot(uri))
+                ?? (try? DocumentSnapshot(url, language: .swift))
+            else {
+              logger.error("Failed to get document snapshot for \(uri.forLogging)")
+              return nil
+            }
+            do {
+              let edits = try await languageService.editsToRename(
+                locations: renameLocations,
+                in: snapshot,
+                oldName: oldName,
+                newName: request.newName
+              )
+              return (uri, edits)
+            } catch {
+              logger.error("Failed to get edits for \(uri.forLogging): \(error.forLogging)")
+              return nil
+            }
+          }
+        }
+        for await case let (uri, textEdits)? in taskGroup where !textEdits.isEmpty {
+          precondition(edits.changes![uri] == nil, "We should create tasks for URIs that already have edits")
```

Comment:
`should not`? ie. `Task was created for URI that already has edits`

But also, we have a check for it above so I'm not sure how useful this actually is.